### ISSUE TITLE
Why does the CApplication error handler logging mechanism skip the first 3 entries of the debug backtrace?

### ISSUE THREAD

**Issue by georaldc** [yiisoft/yii#4355]
https://github.com/yiisoft/yii/blob/d7cf08db201ded76622b4d10c45a25194ebe9d32/framework/base/CApplication.php#L805

I've been noticing my logged PHP errors seemed to be missing a function call step in the beginning and came across the above. This same behavior happens inside the CErrorHandler handleError method but it seems to be appropriate there since the first 3 stacks are indeed not part of when the error first occurs, but I'm not sure if about the CApplication one. It seems like it should only skip the first entry in this case?

**Comment by marcovtwout** [comment]
Skipping entries was added to CErrorHandler since the beginning of time: https://github.com/yiisoft/yii/blame/8d4d1d39cedaf3dd4df5ba2db7f4437f4b9e8f97/framework/core/CErrorHandler.php
And later added to CApplication in 2009 in: https://github.com/yiisoft/yii/commit/67d50c77f881256b0e05b121dd3052799008db59

CErrorHandler is used to display errors to the user, and skipping the first 3 stacks indeed seems to be correct. (these are: CErrorHandler->handleError(), CErrorHandler()->handle() and CApplication->handleError()).

CApplication->handleError() is the registered error handler through PHP's set_error_handler(). It looks like you are correct, only the first entry on the stack should be discarded. Based on the commit history it looks to me like skipping 3 entries was simply the result of a copy/paste. Here is the Yii2 equivalent that only removes 1 entry: https://github.com/yiisoft/yii2/blob/d60252f051eb051dde49052a00a5c7d24981887c/framework/base/ErrorHandler.php#L242

So in short: errors shown to the user have a complete callstack, but errors directed to Yii::log() are missing two entries. Thanks for reporting! I'll prepare a PR.

**Comment by marcovtwout** [comment]
I found one more case where too many entries are skipped in `displayError()`:

Function CApplication->displayError() always skips 3 entries. Function CConsoleApplication::displayError() always skips 4 entries. Added in 2010 in https://github.com/yiisoft/yii/commit/bbfd0eeff31ffb81bd0f67c62fd805b3e7031490

There are many different scenarios and callers:
- Before application->errorHandler is registered:
    - CApplication->handleError() - should skip 2
- After application->errorHandler is registered:
    - Only as console application, through CErrorHandler->handleError() - should skip 4
    - Only as web application, through CErrorHandler->handleError() and then CErrorHandler->renderError() - should skip 5
    - Through CErrorHandler->renderException() and then CErrorHandler->renderError() - should skip 6?

The current number of skipped entries is inconsistent at best, and before the ErrorHandler is registered 1 or 2 too many entries are skipped. I suggest to set both implementations to only skip the minimum number of entries (2).

**Comment by marcovtwout** [comment]
Fixed with https://github.com/yiisoft/yii/pull/4356
